crates/tabby-download/src/lib.rs

```diff
@@ -41,49 +90,70 @@ async fn download_model_impl(
                 "Checksum doesn't match for <{}/{}>, re-downloading...",
                 registry.name, name
             );
-            fs::remove_file(&model_path)?;
+
+            fs::remove_dir_all(registry.get_model_store_dir(name))?;
         } else {
             return Ok(());
         }
     }
 
-    let Some(model_url) = model_info
-        .urls
-        .iter()
-        .flatten()
-        .find(|x| select_by_download_host(x))
-    else {
-        return Err(anyhow!("No valid url for model <{}>", model_info.name));
-    };
-
-    // Replace the huggingface.co domain with the mirror host if it is set.
-    let model_url = if let Ok(host) = std::env::var("TABBY_HUGGINGFACE_HOST_OVERRIDE") {
-        model_url.replace("huggingface.co", &host)
-    } else {
-        model_url.to_owned()
-    };
+    let urls = filter_download_address(model_info);
 
-    let strategy = ExponentialBackoff::from_millis(100).map(jitter).take(2);
-    let download_job = Retry::spawn(strategy, || {
-        download_file(&model_url, model_path.as_path(), &model_info.sha256)
-    });
-    download_job.await?;
-    Ok(())
-}
+    if urls.is_empty() {
+        bail!(
+            "No download URLs available for <{}/{}>",
+            registry.name,
+            model_info.name
+        );
+    }
 
-async fn download_file(url: &str, path: &Path, expected_sha256: &str) -> Result<()> {
-    let dir = path
-        .parent()
-        .ok_or_else(|| anyhow!("Must not be in root directory"))?;
+    // prepare for download
+    let dir = registry.get_model_store_dir(name);
     fs::create_dir_all(dir)?;
 
-    let filename = path
-        .to_str()
-        .ok_or_else(|| anyhow!("Could not convert filename to UTF-8"))?;
-    let intermediate_filename = filename.to_owned() + ".tmp";
+    let mut download_tasks = vec![];
+    for (index, url) in urls.iter().enumerate() {
```

Comment:
The order of the partitioned urls matters. if we name models ourselves, we must sort the URLs first to prevent users from filling in the wrong order.

Reply:
users may provide some malformed URLs, I would prefer that we reserve the order and require all provided URLs in a expected order